content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

```diff
@@ -0,0 +1,204 @@
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot02.jpg)
+
+Looking at one of the certificates eg. apiserver, you can see that it has a list of DNS names and IP addresses. If you want to make this cluster accessible outside, then just describe the additional DNS names in the values file and update the release. This will update the certificate resource, and cert-manager will regeneraate the certificate. You'll no longer need to think about this. If kubeadm certificates need to be renewed at least once a year, here the cert-manager will take care and automatically renew them.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot03.jpg)
+
+Now let's log into the admin container and look at our cluster and nodes. Of course, there are no nodes, yet, because at the moment we have deployed just the control-plane for Kubernetes. But in kube-system namespace we can see some coredns pods and configmaps already appeared. That is, we can conclude that our cluster is working:
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot04.jpg)
+
+Here is the [diagram of the deployed cluster](https://kvaps.github.io/images/posts/Kubernetes-in-Kubernetes-%D0%B8-%D1%84%D0%B5%D1%80%D0%BC%D0%B0-%D1%81%D0%B5%D1%80%D0%B2%D0%B5%D1%80%D0%BE%D0%B2-%D1%81-%D0%B7%D0%B0%D0%B3%D1%80%D1%83%D0%B7%D0%BA%D0%BE%D0%B9-%D0%BF%D0%BE-PXE/Argo_CD_kink_network.html). You can see services for all Kubernetes components: **apiserver**, **controller-manager**, **etcd-cluster** and **scheduler**. And the pods on right side to which they forward traffic.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/argocd01.jpg)
+
+*By the way, the diagram, is drawn in [ArgoCD](https://argoproj.github.io/argo-cd/) — the GitOps tool we use to manage our clusters, and cool diagrams are one of its features.*
+
+### Orchestrating physical servers
+
+OK, now we have a Kubernetes control-plane deployed, but what about worker nodes, how are we going to add them? As I already said, all our servers are bare metal. We do not use virtualization to run Kubernetes, but we orchestrate all physical servers by ourselves.
+
+Also we do use Linux network boot feature very actively. Moreover, this is exactly the booting, not some kind of automation of the installation. When the nodes are booting, they just run a ready-made system image for them. That is, to update any node, we just need to reboot it - and it will download a new image. It is very easy, simple and convenient.
```

Comment:
```suggestion
Also, we do use Linux network boot feature very actively. Moreover, this is exactly the booting, not some kind of automation of the installation. When the nodes are booting, they just run a ready-made system image for them. That is, to update any node, we just need to reboot it - and it will download a new image. It is very easy, simple and convenient.
```